dashboard/app/models/script.rb

```diff
@@ -24,6 +24,7 @@
 
 require 'cdo/script_constants'
```

Comment:
Overall I think I share your impression that none of these things individually are that difficult, or require a change in approach. I just wanted to call out things I did know of, since each thing does add some development cost.

On PLC - yea, I think it should be easy if we want to continue using `generate_plc_objects` as is, though I haven't looked into how much that adds to the seeding time. Both options should work - either continuing to use it, or adding bulk import steps for the PLC objects.

On variants - I don't really know what variants are, haha. It'd help if you could explain this a bit. I just know that I didn't add any logic for them, and it seemed like scripts with variants added more ScriptLevels than they should. But I could believe it's not hard to fix.

On LCDs - this is currently its own model / table, not a field on Level. In schema.rb, it doesn't look like any other table references it, so we could move it to be a field on level. If we did that there'd be no additional work needed. Otherwise, we'd have to add serialization and a bulk import step for these. Neither should be too difficult.

> this (deletion) does not seem like a requirement for shipping this change, because this behavior does not exist in our current system.

I thought this actually did work already for some of the objects in `has_many` associations? But maybe it just deparents them and the old objects stay in the DB, I didn't dig into this.

Reply:
> On variants - I don't really know what variants are, haha. It'd help if you could explain this a bit. I just know that I didn't add any logic for them, and it seemed like scripts with variants added more ScriptLevels than they should. But I could believe it's not hard to fix.

adding a bit more formatting than usual so we can find all these comments later:

### Variants

Variants have two use cases:

1. there is a problem with a level. we want to replace it with a new level in a live/stable script, but we don't want users who already attempted the level to lose their progress. this is sometimes called "level swapping". Example:

before the swap, the script level would be defined like this: https://github.com/code-dot-org/code-dot-org/blob/538a3e2049860ffb4cbefd29dd73d2cfb8caad11/dashboard/config/scripts/coursee.script#L9
https://github.com/code-dot-org/code-dot-

when the level swap is done by adding variants, the old level is marked as inactive and the new level is marked as active. The script level is now defined like this: https://github.com/code-dot-org/code-dot-org/blob/25b5093fd8ad5b1201dbbefc38dc9bf6a1878436/dashboard/config/scripts/coursee-2017.script#L110-L113

when variants are present, the _inactive_ variants have information stored about them on the script level object. it looks like this:
```
[development] dashboard > s = Script.find_by_name('coursee-2017')
[development] dashboard > l = Level.find_by_name('courseE_maze_ramp1')
[development] dashboard > sl = l.script_levels.find{|sl| sl.script == s}
[development] dashboard > puts JSON.pretty_generate sl.properties
{
  "variants": {
    "courseE_maze_ramp1": {
      "active": false
    }
  }
}
```
the logic for which level gets displayed is here: https://github.com/code-dot-org/code-dot-org/blob/77ad50c40eac926a099c3222259f71416e8e6bec/dashboard/app/controllers/script_levels_controller.rb#L421-L426

2. on a certain script level, we want some users to see one level and other users to see another level, depending on whether they are in an experiment. These are typically set up as you see them from day 1 as an intentional experiment, NOT an attempt to fix or change an existing level. Unexpectedly, `active` defaults to `false` when an experiment is present:

* https://github.com/code-dot-org/code-dot-org/blob/c5edef78cc1e9f081bbe7a0e4879692b61b386ce/dashboard/config/scripts/allthethings.script#L219-L222

```
{
  "variants": {
    "csp_socialBelonging_intervention": {
      "active": false,
      "experiments": [
        "socialBelongingTest"
      ]
    }
  }
}
```
The logic for choosing a level to show the user is here: https://github.com/code-dot-org/code-dot-org/blob/77ad50c40eac926a099c3222259f71416e8e6bec/dashboard/app/controllers/script_levels_controller.rb#L428-L434

The logic for converting script DSL into script level properties is here: https://github.com/code-dot-org/code-dot-org/blob/969fa9c4d0da4c829aedaf1e59889134448f450d/dashboard/app/dsl/script_dsl.rb#L225-L235

Reply:
So to explain my earlier comment:

>> variants

> This seems easy, if the variants fields (active and experiments) are going to be stored in the script_level properties, like they are today. Are you considering storing these within levels_script_levels instead? it might be the most logical place, but might also be more work.

I think the easiest way to handle variants for this new seeding approach is probably to leave them as properties on the script level. it seems like this would follow your existing strategy of generally capturing objects as they are with minimal modification. In the future, if we changed the implementation of variants to store the information on level_group_levels, there would be little to no work needed on the seeding code to make that change work.

Reply:
### deletion

>>this (deletion) does not seem like a requirement for shipping this change, because this behavior does not exist in our current system.

>I thought this actually did work already for some of the objects in has_many associations? But maybe it just deparents them and the old objects stay in the DB, I didn't dig into this.

Great point! That is correct, we do not handle deleting of UnitGroup (formerly Course), Script or Level objects, but we do handle deletion of things within a Script such as LessonGroup, Lesson, and ScriptLevel. we typically get this behavior when we assign to a has_many association:
* https://github.com/code-dot-org/code-dot-org/blob/1e21440030e228afcce54177871367c512e74dd0/dashboard/app/models/script.rb#L941
* https://github.com/code-dot-org/code-dot-org/blob/79ff09e81f6dfb2267d77b06106327665bc05fb6/dashboard/app/models/lesson_group.rb#L91
* https://github.com/code-dot-org/code-dot-org/blob/f25f1bac1dce2c68bb58eb1cf712cf62fca39825/dashboard/app/models/lesson.rb#L81

I would say that continuing to delete these (LessonGroup, Lesson, ScriptLevel) is pretty important as failing to do so is likely to cause a regression somewhere. I'll try to share any ideas for how to do this inline.

Reply:
> I would say that continuing to delete these (LessonGroup, Lesson, ScriptLevel) is pretty important as failing to do so is likely to cause a regression somewhere. I'll try to share any ideas for how to do this inline.

@davidsbailey I haven't tried to implement this yet, but I'm reasonably convinced it would work:

Generate a random GUID to use as the "seeding transaction ID". Write this in the attributes for all objects created / updated during this call to `seed_from_json_file`. Then, at the end, query for all objects which don't have that "seeding transaction ID" and delete them.

That's the simplest approach I can think of and should work the same for all models, at the cost of storing some otherwise useless data.

Another approach that should work is to define a `seeding_key` method which returns the uniquely identifying key for each model, and use that to find all objects in the DB which were not created / updated from the serialized data. I'm working on a refactor to use `seeding_key` already, since it might simplify the existing code, so we can see if that seems feasible after that's done.

Reply:
Updated with a different approach which avoids this extra GUID. This approach uses the `seeding_key` for each object to determine whether it was in the list of objects just imported, and if not, destroys them.